MCUmgr  and SMP Client support

jheiskan81 opened the pull request:
### MCUmgr client

MCUmgr client basic implementation for support Image and OS group commands.

Image Group:
    * Image state read/write
    * Image Upload secondary slot
    * Image Erase secondary slot

OS group:
    * Reset
    * Echo service, disabled by default

Operation's are blocked call and can't call inside worker queue.
IMG and OS needs SMP client object for transport.
    
### SMP update

SMP client:
* SMP buffer allocation
* SMP object init for selected SMP transport
* SMP client support for generate request and handling response message.

SMP transport:

Updated SMP transport for trigger client packets.
API for register SMP transport object.

### Unit test

Unit test for MCUmgr & SMP client.

Commit:
mgmt: smp: SMP Client enabler

SMP client support for generate request and handling
response message.

Updated SMP transport for send request.

Added API for register SMP transport.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@SeppoTakalo Please review

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,33 @@
+extern "C" {
+#endif
+
+typedef int (*smp_client_res_fn)(struct net_buf *nb, void *user_data);
```
Files in include/* need documentation in doxygen/sphinx style

nordicjm commented on include/zephyr/mgmt/mcumgr/transport/smp.h:
```diff
@@ -178,6 +180,8 @@ void zephyr_smp_transport_init(struct zephyr_smp_transport *smpt,
 			   (smp_transport_ud_free_fn)ud_free_func, NULL);
 }
 
+struct smp_transport *smp_transport_get(void);
```
As above

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/mcumgr_client.h:
```diff
@@ -0,0 +1,95 @@
+extern "C" {
+#endif
+
+enum mcumgr_image_rsp_t {
```
Doxygen missing for `mcumgr_image_rsp_t`

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/mcumgr_client.h:
```diff
@@ -0,0 +1,95 @@
+	MGMT_IMAGE_ERASE_RSP
+};
+
+enum mcumgr_os_rsp_t {
```
Doxygen missing for `mcumgr_os_rsp_t`

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/mcumgr_client.h:
All high level Doxygen blocks seem to be missing in `mcumgr_client.h`

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/transport/smp.h:
```diff
@@ -109,6 +109,8 @@ struct smp_transport {
 	/* FIFO containing incoming requests to be processed. */
 	struct k_fifo fifo;
+	/* FIFO for client TX queue */
+	struct k_fifo tx_fifo;
```
Should this be under `#ifdef CONFIG_MCUMGR_CLIENT` or SMP client

SeppoTakalo reviewed (changes requested):
Left few remarks

SeppoTakalo commented:
Now MCUMGR client and its OS/Image operations are bundled under one module.

I wonder if it makes sense to have the generic McuMGR as a single file, then optionally enable OS, Image or some other functionality.

jheiskan81 commented:
> Now MCUMGR client and its OS/Image operations are bundled under one module.
> 
> I wonder if it makes sense to have the generic McuMGR as a single file, then optionally enable OS, Image or some other functionality.

OS Reset and Echo are used now. RESET is mandatory. I can replace echo command from Recovery mode by Read Image list command.

de-nordic commented on subsys/mgmt/mcumgr/smp/src/mcumgr_client.c:
```diff
@@ -0,0 +1,713 @@
+			memcpy(src_image->hash, image_data->hash.value, image_data->hash.len);
+			src_image->hash[image_data->hash.len] = '\0';
+			break;
+		case 4:
+			src_image->bootable = map->found ? image_data->bootable : false;
+			break;
+		case 5:
+			src_image->pending = map->found ? image_data->pending : false;
+			break;
+		case 6:
+			src_image->confirmed = map->found ? image_data->confirmed : false;
+			break;
+		case 7:
+			src_image->active = map->found ? image_data->active : false;
+			break;
+		case 8:
+			src_image->permanent = map->found ? image_data->permanent : false;
+			break;
+		}
+		map++;
```
I think this is al little bit too complicated. Take look at `ZCBOR_MAP_DECODE_KEY_VAL`: it takes `zcbor_decoder_t ` signature type function and last parameter is destination for decoded value.
If you define function like
```
bool get_confirmed(zcbor_state_t *state, struct image_data *out)
{
    bool val;
    bool fin = zcbor_bool_decode(state, &val);
    if (!fin) {
       return false;
    }
    // here goes val validation and assignment to image_data 
   ...
    return true;
}
```

you can avoid  looping over the map to validate data, you can validate it as it comes. You can even break out from parsing loop if you return false here. Unfortunately you will have to store some additional failure reason somewhere within your `*out` pointed data structure, but that makes sense because you do not want to mix bulk decoder and validation errors.

jheiskan81 replied:
Good point. I like this idea.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c:
```diff
@@ -226,21 +229,25 @@ img_mgmt_state_read(struct smp_streamer *ctxt)
-		       zcbor_int32_put(zse, i >> 1)))				&&
-		     zcbor_tstr_put_lit(zse, "slot")				&&
-		     zcbor_int32_put(zse, i % 2)				&&
+		ok = zcbor_map_start_encode(zse, MAX_IMG_CHARACTERISTICS) &&
+		     (CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER == 1 ||
+		      (zcbor_tstr_put_lit(zse, "image") && zcbor_int32_put(zse, i >> 1))) &&
+		     zcbor_tstr_put_lit(zse, "slot") && zcbor_int32_put(zse, i % 2) &&
```
I think original code was not checked return values at all

jheiskan81 replied:
Ah I can put this to original. It is just code formatting change.

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/mcumgr_client.c:
```diff
@@ -0,0 +1,713 @@
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h>
+#include <mgmt/mcumgr/grp/img_mgmt/img_mgmt_priv.h>
+
+#include <assert.h>
```
Copy paste I will remove

jheiskan81 commented:
@SeppoTakalo and @nordicjm I removed Canonical fix commit. I have updated SMP and MCUMGR client commits missing headers doxygen comment's.

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/mcumgr_client.h:
```diff
@@ -0,0 +1,211 @@
+struct mcumgr_image_data {
+	int slot_num;
+	int img_num;
+	char hash[33];
```
there is a define somewhere in mcumgr for the sha256 hash size, and it should be 32 bytes, not 33 (null character isn't needed)

jheiskan81 replied:
Ok. Let me think. Hash could be also shorter than 32 in some point. That's why 1 extra byte is needed for max lenght

nordicjm replied:
Hash is either 0 if not present or 32 for an sha256 - other lengths are invalid

jheiskan81 replied:
I agree.

nordicjm commented on subsys/mgmt/mcumgr/smp/src/mcumgr_client.c:
```diff
@@ -0,0 +1,756 @@
+
+static bool get_image_version(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	struct zcbor_string val = {0};
+	bool fin = zcbor_tstr_decode(state, &val);
+
+	if (!fin || !val.len) {
+		return false;
+	}
+
+	if (val.len >= sizeof(out->version)) {
+		val.len = sizeof(out->version) - 1;
+	}
+	memcpy(out->version, val.value, val.len);
+	out->version[val.len] = '\0';
+	return true;
+}
+
+static bool get_image_hash(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	struct zcbor_string val = {0};
+	bool fin = zcbor_bstr_decode(state, &val);
+
+	if (!fin || !val.len) {
+		return false;
+	}
+
+	if (val.len >= sizeof(out->hash)) {
+		val.len = sizeof(out->hash) - 1;
+	}
+	memcpy(out->hash, val.value, val.len);
+	out->hash[val.len] = '\0';
+	return true;
```
Not really sure why any of this is needed when you can simply have the zcbor object itself and use it directly which avoids duplicating data because it doesn't copy the contents anywhere

jheiskan81 replied:
Now data is not duplicated. Code copy from message to image data buffer.

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/mcumgr_client.h:
```diff
@@ -0,0 +1,211 @@
+ */
+enum mcumgr_os_rsp_t {
+	/** OS response */
+	MGMT_OS_ECHO_RSP = 0,
```
All of this needs splitting up, a "client" should only have the functionality needed for the client, i.e. no command processing, command processing should be moved to independent files (for each group e.g. fs, os, img, etc.)

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/mcumgr_client.h:
```diff
@@ -0,0 +1,211 @@
+	int status;
+	union {
+		struct mcumgr_image_state image_state;
+		size_t image_uppload_offset;
```
*upload

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/mcumgr_client.h:
```diff
@@ -0,0 +1,211 @@
+};
+
+/**
+ * @brief MCUMGR client IMAGE group response callback.
```
*MCUmgr

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,282 @@
+};
+
+static struct smp_client_cmd_req smp_cmd_req_bufs[SMP_CMD_MAX];
+static struct smp_client_data_base smp_client_data;
```
A single client, and with global file scope? That does not seem very extensible.

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,282 @@
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&smp_client_data.cmd_list, entry, tmp, node) {
+		if (smpt && k_uptime_get() - entry->timestamp < 1000) {
+			continue;
+		} else if (smpt && entry->retry_cnt < 15) {
```
15 retries is a lot - should be user configurable

jheiskan81 replied:
Erase command may take a more time. I can put smaller than 15

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,282 @@
+
+	if (!sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* Re-schedule new timeout check*/
+		k_work_reschedule(&smp_client_data.work_delay, K_MSEC(50));
```
If this does a resend every 1 second, why is it being ran every 50ms? Maybe store the next timeout and schedule it for that amount of time.

jheiskan81 replied:
Good Point. Let me re-think this.

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,282 @@
+{
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* Enable timer */
+		k_work_reschedule(&smp_client_data.work_delay, K_MSEC(100));
```
100ms?

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,279 @@
+	}
+}
+
+static struct smp_client_cmd_req *smp_client_response_discover(const struct smp_hdr *res_hdr)
+{
+	struct smp_hdr smp_header;
+	enum mcumgr_op_t response;
+	struct smp_client_cmd_req *cmd_req;
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		return NULL;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&smp_client_data.cmd_list, cmd_req, node) {
+		smp_read_hdr(cmd_req->nb, &smp_header);
+		if (smp_header.nh_op == MGMT_OP_READ) {
+			response = MGMT_OP_READ_RSP;
+		} else {
+			response = MGMT_OP_WRITE_RSP;
+		}
+
+		if (smp_header.nh_seq != res_hdr->nh_seq) {
+			continue;
+		} else if (res_hdr->nh_op != response) {
+			continue;
+		}
+
+		return cmd_req;
+	}
+	return NULL;
+}
```
Not really sure what this is doing?

jheiskan81 replied:
It will discover proper request entry for received message

de-nordic commented on subsys/mgmt/mcumgr/smp/src/mcumgr_client.c:
```diff
@@ -0,0 +1,756 @@
+
+	struct zcbor_map_decode_key_val upload_res_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_DECODER("rc", zcbor_int32_decode, &response.status),
+		ZCBOR_MAP_DECODE_KEY_DECODER("rsn", zcbor_tstr_decode, &rsn)};
```
You are not using "rsn" for anything, you may leave it out.

de-nordic commented on subsys/mgmt/mcumgr/smp/src/smp.c:
```diff
@@ -318,24 +321,37 @@ int smp_process_request_packet(struct smp_streamer *streamer, void *vreq)
+			/* Send the response. */
+			rc = streamer->smpt->output(rsp);
+			rsp = NULL;
+		} else if (req_hdr.nh_op == MGMT_OP_READ_RSP ||
```
You can add `IS_ENABLED(CONFIG_SMP_CLIENT) && ` to if and in such case, you can avoid #ifdef below.

de-nordic commented on subsys/mgmt/mcumgr/smp/src/mcumgr_client.c:
```diff
@@ -0,0 +1,756 @@
+	return true;
+}
+
+static bool get_image_slot(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	bool fin = zcbor_int32_decode(state, &out->slot_num);
+
+	return fin;
+}
+
+static bool get_image_num(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	bool fin = zcbor_int32_decode(state, &out->img_num);
+
+	return fin;
+}
```
I do not get those. The functions here are very simple, the validation code previously was quite more complicated, is there some logic missing yet? Otherwise why not directly call the `zcbor_` functions.

nordicjm commented on subsys/mgmt/mcumgr/smp/src/mcumgr_client.c:
```diff
@@ -0,0 +1,756 @@
+	return fin;
+}
+
+static bool get_image_num(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	bool fin = zcbor_int32_decode(state, &out->img_num);
+
+	return fin;
+}
+
+static bool get_image_bootable(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	bool fin = zcbor_bool_decode(state, &out->bootable);
+
+	return fin;
+}
+
+static bool get_image_pending(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	bool fin = zcbor_bool_decode(state, &out->pending);
+
+	return fin;
+}
+
+static bool get_image_confirmed(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	bool fin = zcbor_bool_decode(state, &out->confirmed);
+
+	return fin;
+}
+
+static bool get_image_active(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	bool fin = zcbor_bool_decode(state, &out->active);
+
+	return fin;
+}
+
+static bool get_image_pernament(zcbor_state_t *state, struct mcumgr_image_data *out)
```
It should be storing the data *locally* so it can be verified before putting it in a global or shared scope - see https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt.c#L391 it should be working similarly to that

jheiskan81 replied:
OK. let me put back those local buffer's and copy those to image data list.

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/mcumgr_client.c:
```diff
@@ -0,0 +1,756 @@
+			goto out;
+		}
+		/* Check that mandatory parametrs have decoded */
+		if (!list_res_decode[0].found || !list_res_decode[1].found ||
+		    !list_res_decode[2].found) {
```
I agree that I don't like to use that .found. I can add a mask for those mandatory parameter's.

jheiskan81 commented:
I have been fixed now raised item's:
SMP:
SMP CLIENT retry logic is updated now more dynamic and retry shedule is using exact timestamp to next retry send
MCUmgr client:
Updated Image group state read response IMAGE list parser.
Removed rsn unused caesd
Thanks for good feedback @nordicjm  and @de-nordic.

jheiskan81 commented:
MCUmgr client file can be splitted to own seprated groups file but I want to review main functionality first.

jheiskan81 commented:
rebased with upstream

jheiskan81 commented:
@SeppoTakalo @nordicjm @de-nordic Please review changes.

jheiskan81 commented:
Splitted mcumgr_client to own group's and defined own header files.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,558 @@
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+	if (value.len != 6 || memcmp(value.value, smp_images_str, 6)) {
```
What does 6 refer to?

jheiskan81 replied:
Ah good point It is string lenght without \0. I will do define for that.

jheiskan81 commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,149 @@
+ * @brief MCUmgr Image Group response types.
+ */
+enum mcumgr_image_rsp_t {
+	/** Image state write or read response */
+	MGMT_IMAGE_STATE_RSP = 0,
+
+	/** Image Upload response*/
+	MGMT_IMAGE_UPLOAD_RSP,
+
+	/** Secondary image erase response */
+	MGMT_IMAGE_ERASE_RSP
```
OK. I got your point.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,558 @@
+	int upload_status;
+};
+
+struct image_read_response {
```
OK let me check this possibility

jheiskan81 replied:
I think I remove structure and use parameter defined 1 by 1. It is best solution

jheiskan81 commented:
@nordicjm I have fix next items:
- removed those extra structure and enumeration updates. 
- added Kconfig for SMP CLIENT retry and timeout.
- Fix SMP_CLIENT auto select by depends on

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+		image_info[image_count].img_num = img_num;
+		image_info[image_count].slot_num = slot_num;
+		memcpy(image_info[image_count].hash, hash.value, IMG_MGMT_HASH_LEN);
+		if (version.len >= IMG_MGMT_VER_MAX_STR_LEN) {
```
`>=` doesn't seem right if you are setting it to this value

jheiskan81 replied:
> is correct

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+ */
+#define CBOR_ENCODER_STATE_NUM 3
+
+#define MCUMGR_IMAGE_LIST_SIZE 2
```
Should be user configurable

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+
+	ok = zcbor_tstr_decode(zsd, &value);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+	if (value.len != IMAGES_STR_LEN || memcmp(value.value, smp_images_str, IMAGES_STR_LEN)) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
+		goto out;
+	}
+
+	ok = zcbor_list_start_decode(zsd);
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		rc = MGMT_ERR_EUNKNOWN;
```
Not really sure why the response status and rc values differ?

jheiskan81 replied:
Thats a goop point. I think I copy idea some where but now It looks a weird for me also. Let me double check if I use same for return and notification.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+				break;
+			}
+			LOG_ERR("Corrupted Image data %d", rc);
+			response.status = MGMT_ERR_ECORRUPT;
+			rc = MGMT_ERR_EINVAL;
```
Is it corrupt or an invalid value?

jheiskan81 replied:
I start Think here that it should use MGMT_ERR_EINVAL to status also. Is it make sense @nordicjm ?

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+
+static struct upload_progress upload_state;
+static mgmt_client_gr_image_fn client_gr_image_fn;
+static struct mcumgr_image_data image_info[MCUMGR_IMAGE_LIST_SIZE];
```
This is not multiple transport safe (does not need changing for first implementation, just a note)

jheiskan81 replied:
I will replace this in callback remove update state. User give buffer in future

jheiskan81 replied:
Static buffer removed user gives buffer pointer for storage image list data.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+		ZCBOR_MAP_DECODE_KEY_DECODER("rc", zcbor_int32_decode, &response.status)};
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+		upload_state.upload_status = MGMT_ERR_ETIMEOUT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return 0;
```
There's 3 errors here, timed out twice (does it need to be set twice) and 0 for OK?

jheiskan81 replied:
0 is OK because SMP client indicate timeout for user. I need to check Why I set upload_state.upload_status.

jheiskan81 replied:
I will return timeout in that case also. It not break anything

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+	if (rc || response.status == -1 || response.image_upload_offset == SIZE_MAX) {
+		response.status = MGMT_ERR_ECORRUPT;
+		upload_state.upload_status = MGMT_ERR_ECORRUPT;
+		client_gr_image_fn(&response);
+		k_sem_give(user_data);
+		return MGMT_ERR_EINVAL;
```
Instead of doing this a lot, have a `goto` e.g. `goto end;` then have the `end:` part where it gives the semaphore and returns e.g. `rc`

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+	while (upload_state.request_length != upload_state.writed_length) {
+		write_ptr = data + upload_state.writed_length;
+		write_length = upload_state.request_length - upload_state.writed_length;
+		if (write_length > 256) {
+			write_length = 256;
```
Why 256?

jheiskan81 replied:
That was starting point value. I need to change this based on net_buf size.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+					       MGMT_OP_WRITE);
+		if (!nb) {
+			k_mutex_unlock(&mcumgr_mutex);
+			return -ENOMEM;
```
Should this be returning an error code or an mcumgr status code? The 2 are not the same

jheiskan81 replied:
Thats true.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,542 @@
+			     zcbor_size_put(zse, upload_state.image_size) &&
+			     zcbor_tstr_put_lit(zse, "off") &&
+			     zcbor_size_put(zse, upload_state.offset) &&
+			     zcbor_tstr_put_lit(zse, "sha") && zcbor_bstr_put_lit(zse, "12345") &&
```
I will generate this dynamically for new started upload. Need to think about what data i use or just randomize.

jheiskan81 replied:
Hash from whole file in beginning is impossible if I just download that piece of piece. I can take sha from first part of image

nordicjm replied:
> Hash from whole file in beginning is impossible if I just download that piece of piece. I can take sha from first part of image

No, this is not acceptable, it is the full sha256 hash of the whole of the image. This is clearly listed in the specification and on the documentation.

jheiskan81 replied:
I can't then give a hash if I download a image 1024 bytes slots.

nordicjm replied:
But why would you be downloading in 1024 byte slots? This is being transferred from mcuboot to mcuboot, correct? We completely ruled out transferring from one device to the other and trying to implement this over hci.

jheiskan81 replied:
LwM2M Firmware object use download client for download image from server. Normally DFU MCUBOOT target write images to flash secondary slot direct.
SMP Client -> mcuboot
When we have SMP instances we use new DFU SMP target which download same why but DFU target write use MCUmgr IMG groupr + SMP Upload command to forward image pieces to external MCUBoot.

jheiskan81 replied:
HASH is now defined by MCUmgr user.

jheiskan81 commented:
@nordicjm I have been gone trough your input and updated commit's. I update API for supporting to Set HASH for upload or not. Client dynamically support both now. I have been removed MCUmgr client callback API for request.

jheiskan81 commented:
Rebased with upstream and updated Upload data write offset response handle for last session continue state

jheiskan81 commented:
@SeppoTakalo @nordicjm @de-nordic Please review.

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,172 @@
+
+#ifdef CONFIG_MCUMGR_GRP_OS_ECHO_REQ
+
+static const char *echo_cmp_string;
+static int echo_status;
+/* Mimimal number of ZCBOR encoder states to provide full encoder functionality plus one for backup
+ */
+#define CBOR_ENCODER_OS_STATE_NUM 3
```
put all these at the top of the file with other variable/define declarations (copy the #ifdef up there)

jheiskan81 replied:
This file is only builded when CONFIG_MCUMGR_CLIENT_GRP_OS is defined. so up of the src file not need #ifdef CONFIG_MCUMGR_GRP_OS_ECHO_REQ

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,172 @@
+		goto end;
+	}
+
+	if (val.len != strlen(echo_cmp_string)) {
+		echo_status = MGMT_ERR_EINVAL;
```
Not necessarily a failure, the other side of the communication might have different buffer sizes and might truncate the response, which isn't an error in itself.

jheiskan81 replied:
Ah ok good point I did not think that was possible.

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,172 @@
+	zcbor_state_t zse[CBOR_ENCODER_OS_STATE_NUM];
+
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+	/* allocate buffer */
```
Comment not needed

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,295 @@
+	smp_client_res_fn cb;
+	void *user_data;
+
+	/* discover request for coming response*/
```
```suggestion
	/* Discover request for incoming response */
```

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,295 @@
+
+void smp_client_buf_free(struct net_buf *nb)
+{
+	while (nb->ref) {
+		smp_packet_free(nb);
+	}
```
While? Surely the number of refs should be known and counted properly so it only needs to be free'd once for each time it has been ref'd?

jheiskan81 replied:
Removed while.

nordicjm commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,295 @@
+
+	smpt = smp_transport_get();
+	if (!smpt) {
+		return MGMT_ERR_EACCESSDENIED;
```
I don't think this really represents the error condition

nordicjm commented on subsys/mgmt/mcumgr/transport/include/mgmt/mcumgr/transport/smp_internal.h:
```diff
@@ -45,6 +45,19 @@ struct zephyr_smp_transport;
 
+#ifdef CONFIG_SMP_CLIENT
+/**
+ * @brief Enqueues an client SMP request packet to Transmit.
```
```suggestion
 * @brief Enqueues a client SMP request packet for transmission.
```

nordicjm commented on subsys/mgmt/mcumgr/transport/include/mgmt/mcumgr/transport/smp_internal.h:
```diff
@@ -45,6 +45,19 @@ struct zephyr_smp_transport;
+ *
+ * This function always consumes the supplied net_buf.
+ *
+ * @param smtp                  The transport to use to send the corresponding
+ *                                  response(s).
+ * @param nb                    The request packet to process.
```
Align properly

nordicjm commented on subsys/mgmt/mcumgr/transport/include/mgmt/mcumgr/transport/smp_internal.h:
```diff
@@ -45,6 +45,19 @@ struct zephyr_smp_transport;
+ *
+ * This function always consumes the supplied net_buf.
+ *
+ * @param smtp                  The transport to use to send the corresponding
```
*smpt, smtp is something very different

jheiskan81 replied:
Ah but RX is use same name already?

nordicjm commented on subsys/mgmt/mcumgr/transport/src/smp.c:
```diff
@@ -157,7 +168,19 @@ smp_transport_init(struct smp_transport *smpt,
+#ifdef CONFIG_SMP_CLIENT
+	k_fifo_init(&smpt->tx_fifo);
+	/* Store Transport pointer */
+	smpt_transport = smpt;
```
So how is this going to work if a device has multiple transports registered?

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,137 @@
+/**
+ * @brief MCUmgr Image list response.
+ *
+ *	Structure image_list_length and image_list pointer are defined by user for store received
+ *	image list data. image_list_length is updated by MCUmgr client for response.
```
```suggestion
 *	Structure image_list_length and image_list pointer are defined by user for storing received
 *	image list data. image_list_length is updated by MCUmgr client for response.
```

And the indentation looks off here

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,137 @@
+ */
+struct mcumgr_image_upload {
+	/** status */
+	int status;
```
Is this an int or `mcumgr_err_t`?

jheiskan81 replied:
mcumgr_err_t is a correct

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,137 @@
+ * @brief Init started Image Upload Process
+ *
+ * @param image_size Size of image in bytes.
+ *
+ * @param image_num Image Num.
+ *
+ * @param image_hash Pointer to HASH for image. Use NULL when HASH from image is not available.
+ *
```
Structure these sections like https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/rtc.h#L206

No empty newline gap between the params, return needs to have different lines for the values and what the conditions are.

Do this for all sections here and below

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,137 @@
+};
+
+/**
+ * @brief Init started Image Upload Process
```
Rephrase this

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @param hash		Pointer to Hash (Needed for test).
+ *
+ * @param hash_len	length of Hash (Needed for test).
```
Not sure we need a hash_len param, either the hash is not provided (hash == NULL) or it should be the size of the define

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @param confirm	Set false for test and true for confirmation.
+ *
+ * @param res_buf	Pointer for store image list res_buf->image_list_length inicate buffer
+ *			size res_buf->image_list. res_buf->image_list_length is updated for parsed
+			image list data.
```
This doesn't seem right

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,137 @@
+/**
+ * @brief Read Image list
+ *
+ * @param res_buf	Pointer for store image list res_buf->image_list_length inicate buffer
+ *			size res_buf->image_list. res_buf->image_list_length is updated for parsed
+			image list data.
```
Fix spelling, grammar and missing *

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,83 @@
+ *
+ * @param nb net_buf for response
+ *
+ * @param user_data same user data returned than for request
```
If you mean the same data, then:
```suggestion
 * @param user_data same user data that was provided as part of the request
```

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,83 @@
+ */
+typedef int (*smp_client_res_fn)(struct net_buf *nb, void *user_data);
+
+
```
replace double newlines with single

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,83 @@
+/**
+ * @brief SMP net buf allocate and SMP header pre-write
+ *
+ * @param group SMP grouo id
```
typo

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,83 @@
+ */
+void smp_client_buf_free(struct net_buf *nb);
+
+
```
As above

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,83 @@
+/**
+ * @brief SMP client data send request
+ *
+ * @param nb net_buf builded SMP packet for send
```
typo

nordicjm commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,83 @@
+ *
+ * @param user_data user defined data pointer which will be returned back to response callback
+ *
+ * @param timeout_in_sec Define How many times re-sended with CONFIG_SMP_CMD_RETRY_TIME period.
```
this comment does not make sense, it's the timeout in seconds, but the comment suggests it's actually not in seconds and is actually the number of times the retry is sent?

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,25 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	bool "SMP Client support"
+	help
+	  This will enable SMP Request generation and response handling.
+if SMP_CLIENT
```
```suggestion

if SMP_CLIENT

```

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,25 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	default 5
+	help
+	  This define how many times Packet is re-sended by using SMP_CMD_RETRY_TIME value.
+config SMP_CMD_RETRY_TIME
```
newline

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,25 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	range 2 30
+	default 5
+	help
+	  This define how many times Packet is re-sended by using SMP_CMD_RETRY_TIME value.
```
typo and reword because as above, the comment doesn't make sense

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,25 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	  This will enable SMP Request generation and response handling.
+if SMP_CLIENT
+config SMP_CMD_DEFAULT_LIFE_TIME
+	int "SMP command Lifetime in seconds"
```
```suggestion
	int "SMP command lifetime in seconds"
```

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,25 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	help
+	  This define how many times Packet is re-sended by using SMP_CMD_RETRY_TIME value.
+config SMP_CMD_RETRY_TIME
+	int "SMP command re-send periond in ms"
```
```suggestion
	int "SMP command re-send period in ms"
```

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,25 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	default 500
+	help
+	  This will set how long time SMP client wait response before re-send
+endif # SMP_CLIENT
```
missing newline before endif

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,25 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	  This define how many times Packet is re-sended by using SMP_CMD_RETRY_TIME value.
+config SMP_CMD_RETRY_TIME
+	int "SMP command re-send periond in ms"
+	range 100 1000
+	default 500
+	help
+	  This will set how long time SMP client wait response before re-send
```
This values do not make sense, an erase can take upwards of 10 seconds, so why would you resend the command after 0.5s?

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:
```diff
@@ -109,3 +109,9 @@ module-str = mcumgr_grp_img
 endif
+
+config MCUMGR_CLIENT_GRP_IMG
+	bool "Mcumgr client handlers for image management"
```
*MCUmgr

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+#include <pm_config.h>
```
No way

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+	bool bootable, pending, confirmed, active, permanent, ok;
+	size_t decoded;
+	struct zcbor_map_decode_key_val list_res_decode[] = {
+		/** Mandatory components */
+		ZCBOR_MAP_DECODE_KEY_DECODER("version", zcbor_tstr_decode, &version),
+		ZCBOR_MAP_DECODE_KEY_DECODER("hash", zcbor_bstr_decode, &hash),
+		ZCBOR_MAP_DECODE_KEY_DECODER("slot", zcbor_int32_decode, &slot_num),
+		/** End of mandatory components */
```
Don't need /** for comments here because there files are not documented, replace the "end of" with "optional", these aren't components, they're fields

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+			image_info->status = MGMT_ERR_EINVAL;
+			goto out;
+		}
+		/* Check that mandatory parametrs have decoded */
```
typo

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+			ok = zcbor_tstr_put_lit(zse, "off") &&
+			     zcbor_size_put(zse, upload_state.offset);
+		}
+		/* Write hash when it defined and offset is 0*/
```
space before */

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+	image_info = res_buf;
+
+	/* allocate buffer */
```
Comment can go

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+	image_info = res_buf;
+
+	/* allocate buffer */
```
Comment can go. Comments should explain what the point of the code is or why a particular method was chosen, it should not just say what the function call is doing when that's obvious from the name of the function itself

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+
+	/* allocate buffer */
```
Etc. for all

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/Kconfig:
```diff
@@ -173,3 +173,15 @@ module-str = mcumgr_grp_os
+	depends on SMP_CLIENT
+	help
+	  Enables mcumgr client request and response handlers for OS management
+if MCUMGR_CLIENT_GRP_OS
```
```suggestion

if MCUMGR_CLIENT_GRP_OS

```

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/Kconfig:
```diff
@@ -173,3 +173,15 @@ module-str = mcumgr_grp_os
+if MCUMGR_CLIENT_GRP_OS
+	config MCUMGR_GRP_OS_ECHO_REQ
+	bool "Support for echo command request"
+	default n
```
n is the default, line can go

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,172 @@
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+#include <pm_config.h>
```
No

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,172 @@
+#include <mgmt/mcumgr/util/zcbor_bulk.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+
+
```
Don't need double line space for this

nordicjm commented:
In addition there needs to be a simple test for this, because if there was you'd see it would fail to compile for all boards

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+
+	upload_state.request_length = length;
+	upload_state.writed_length = 0;
+	/* Calculate max data length based net buf size with 128 bytes for SMP header + CBOR */
+	max_data_length = CONFIG_MCUMGR_TRANSPORT_NETBUF_SIZE - 128;
```
It is difficult to estimate cbor overhead. Any idea to do that optimal way?

jheiskan81 replied:
I have now created function which calculate cbor overhead size at upload init.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/os_mgmt/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,172 @@
+ */
+
+#define LOG_MODULE_NAME mcumgr_client_os_grp
+#define LOG_LEVEL	CONFIG_LWM2M_LOG_LEVEL
```
Ah this is copy paste issue.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+		} else if (offset_before_send + write_length ==  upload_state.offset) {
+			/* Send operation normal state */
+			upload_state.writed_length += write_length;
+		} else  {
+			/* Stop this packet send offset is at further than this current packet */
+			goto end;
```
Yes I will refactor this part When I was calculating things dynamically i was seeing problem when I was push data lenght which which is not alignment by word size.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,540 @@
+ */
+
+#define LOG_MODULE_NAME mcumgr_client_img_grp
+#define LOG_LEVEL	CONFIG_LWM2M_LOG_LEVEL
```
Copy paste issue line is not needed

Commit:
tests: mcumgr: MCUmgr and smp client unit test

Added unit test for testing IMG and OS group component's.

Added Unit test for SMP Client.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

nordicjm commented on tests/subsys/mgmt/mcumgr/mcumgr_client/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+target_sources(app PRIVATE ${ZEPHYR_BASE}/subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c)
+target_sources(app PRIVATE ${ZEPHYR_BASE}/subsys/mgmt/mcumgr/grp/os_mgmt_client/src/os_mgmt_client.c)
+
+add_compile_definitions(CONFIG_SMP_CLIENT=y)
+add_compile_definitions(CONFIG_SMP_CMD_DEFAULT_LIFE_TIME=5)
+add_compile_definitions(CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER=1)
+add_compile_definitions(CONFIG_MCUMGR_GRP_OS_ECHO_REQ=y)
```
For testing like this, I think this should be designed around the "unit testing" board which excludes most zephyr files - @tejlmand thoughts?

jheiskan81 replied:
I update config that prj.conf enable main setup. I will stub SMP client for unit test for testing MCUmgr client component's

nordicjm commented on subsys/mgmt/mcumgr/transport/src/smp.c:
```diff
@@ -151,10 +162,21 @@ int smp_transport_init(struct smp_transport *smpt)
+#ifdef CONFIG_SMP_CLIENT
+	k_fifo_init(&smpt->tx_fifo);
+	/* Store Transport pointer */
+	smpt_transport = smpt;
```
As mentioned before, this is not acceptable

jheiskan81 replied:
I mention add my comment that this is still open. I have fix coming to this soon

jheiskan81 commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -32,6 +32,30 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+
+config SMP_CMD_RETRY_TIME
+	int "SMP command re-send period in ms"
+	range 100 1000
```
This is 1 retry. PC go app send much aggressive. If we want extend max value I'm fine

jheiskan81 commented on subsys/mgmt/mcumgr/grp/os_mgmt/Kconfig:
```diff
@@ -173,3 +173,17 @@ module-str = mcumgr_grp_os
 
 endif
+
+menuconfig MCUMGR_CLIENT_GRP_OS
+	bool "MCUmgr client request and response handlers for OS management"
+	depends on SMP_CLIENT
+	help
+	  Enables mcumgr client request and response handlers for OS management.
```
ah good point.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/os_mgmt_client/Kconfig:
```diff
@@ -0,0 +1,26 @@
+if MCUMGR_GRP_OS_CLIENT
+
+	config MCUMGR_GRP_OS_ECHO_REQ
+	bool "Support for echo command request"
```
I was following server mode where it was enabled by default.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/os_mgmt_client/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,157 @@
+}
+
+#ifdef CONFIG_MCUMGR_GRP_OS_ECHO_REQ
+static int echo_status;
```
need for this is removed at my refactor when I update Commit's

jheiskan81 commented on include/zephyr/mgmt/mcumgr/transport/smp.h:
```diff
@@ -97,6 +97,16 @@ typedef bool (*smp_transport_query_valid_check_fn)(struct net_buf *nb, void *arg
  */
 struct smp_transport_api_t {
 	/** Transport's send function. */
+	/* Must be the first member. */
+	struct k_work work;
+
+	/* FIFO containing incoming requests to be processed. */
+	struct k_fifo fifo;
+#ifdef CONFIG_SMP_CLIENT
+	/* FIFO for client TX queue */
+	struct k_fifo tx_fifo;
```
That is confict resolve issue I fix. sorry confusion.

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,293 @@
+	header->nh_group = sys_cpu_to_be16(group);
+	header->nh_id = id;
+	header->nh_seq = smp_client_data.smp_seq++;
+}
```
Ah I need to check that

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,595 @@
+	/* Calculate max data length based on net_buf size - (SMP header + CBOR message_len) */
+	max_data_length = CONFIG_MCUMGR_TRANSPORT_NETBUF_SIZE - (upload_state.upload_header_size +
+			  MGMT_HDR_SIZE);
+	/* Check that allways send full of word's */
```
Good point i fix this.

jheiskan81 commented:
@nordicjm @SeppoTakalo  @de-nordic Please review.
Refactored SMP and MCUmg client dependencies:
 
- SMP client use own Object by registered SMP transport.
-  MCUmgr IMG and OS group have a own Client object which are linked to SMP client object and transport.

Upload image length trim use now own kconfig parameter which default value is 4.

Commit:
mgmt: smp: MCUmgr Client support

MCUmgr client basic implementation for support Image and OS grpup
commands.

Image Group:
* Image state read/write
* Image Upload secondary slot
* Image Erase secondary slot

OS group:
* Reset
* Echo service, disabled by default

Opeartion's are blocked call and cant't call inside worker queue.
IMG and OS need to be SMP client object for transport.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h:
```diff
@@ -0,0 +1,63 @@
+ * @brief OS mgmt client object
+ */
+struct os_mgmt_client {
+	/* SMP client object  */
```
```suggestion
	/** SMP client object  */
```

Same in this header. Doxygen blocks start with `/*` instead of `/**`

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h:
```diff
@@ -0,0 +1,63 @@
+};
+
+/**
+ * @brief OS mgmt client object init
```
```suggestion
 * @brief Initialize OS management client.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h:
```diff
@@ -0,0 +1,63 @@
+void os_mgmt_client_init(struct os_mgmt_client *client, struct smp_client_object *smp_client);
+
+/**
+ * @brief SMP Send OS Echo message
```
```suggestion
 * @brief Send SMP message for Echo command
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h:
```diff
@@ -0,0 +1,63 @@
+int os_mgmt_client_echo(struct os_mgmt_client *client, const char *echo_string);
+
+/**
+ * @brief SMP OS Reset command
```
```suggestion
 * @brief Send SMP Reset command.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,101 @@
+#endif
+
+/**
+ * @brief Init a SMP client object.
```
```suggestion
 * @brief Initialize a SMP client object.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,101 @@
+int smp_client_object_init(struct smp_client_object *smp_client, enum smp_trasnport_type smp_type);
+
+/**
+ * @brief SMP client send response callback function
```
```suggestion
 * @brief Response callback for SMP send.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,101 @@
+typedef int (*smp_client_res_fn)(struct net_buf *nb, void *user_data);
+
+/**
+ * @brief SMP client response handler
```
```suggestion
 * @brief SMP client response handler.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,101 @@
+int smp_client_single_response(struct net_buf *nb, const struct smp_hdr *res_hdr);
+
+/**
+ * @brief SMP net buf allocate and SMP header pre-write
```
```suggestion
 * @brief Allocate buffer and initialize with SMP header.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,101 @@
+struct net_buf *smp_client_buf_allocation(struct smp_client_object *smp_client, uint16_t group,
+					  uint8_t command_id, uint8_t op);
+/**
+ * @brief Frees an SMP client net_buf
```
```suggestion
 * @brief Free a SMP client buffer.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/smp/smp_client.h:
```diff
@@ -0,0 +1,101 @@
+void smp_client_buf_free(struct net_buf *nb);
+
+/**
+ * @brief SMP client data send request
```
```suggestion
 * @brief SMP client data send request.
```

SeppoTakalo commented on include/zephyr/mgmt/mcumgr/transport/smp.h:
```diff
@@ -162,6 +181,26 @@ void smp_rx_remove_invalid(struct smp_transport *zst, void *arg);
+int smp_transport_register(struct smp_transport *smpt, enum smp_trasnport_type smpt_type);
+
+/**
+ * @brief Discover a registred  Zephyr SMP transport object for client.
```
```suggestion
 * @brief Discover a registered  SMP transport object.
```

SeppoTakalo commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -33,6 +33,31 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	range 2 30
+	default 5
+	help
+	  This define lifetime for SMP client send request. This configure is used if request
+	  giving timeout is 0.
```
```suggestion
	  This define lifetime for SMP client send request. This configure is used if request
	  with timeout 0 are used.
```

SeppoTakalo reviewed (commented):
Left few remarks regarding the Doxygen documentation.
Otherwise this looks fine to me.

SeppoTakalo reviewed (commented):
Left few remarks regarding the Doxygen documentation.
Otherwise this looks fine to me.

jheiskan81 commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,204 @@
+ * @param image_list pointer to image list
+ *
+ */
+void img_mgmt_client_image_list_print(int image_list_length, struct mcumgr_image_data *image_list);
```
Good point this is only printing at debug purpose. I use internally this so I may move this away from public header.

aunsbjerg commented on include/zephyr/mgmt/mcumgr/transport/smp.h:
```diff
@@ -134,6 +139,20 @@ struct smp_transport {
+/**
+ * @brief SMP transport type for client registration
+ */
+enum smp_trasnport_type {
```
Spelling error, should be `smp_transport_type`

aunsbjerg commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,297 @@
+	header->nh_seq = seq;
+}
+
+static void snmp_client_trasnport_work_fn(struct k_work *work)
```
`trasnport` -> `transport`

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,195 @@
+	bool confirmed: 1;
+	/** Active image */
+	bool active: 1;
+	/** Pernament image state */
```
*Permanent

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,195 @@
+ * @brief Image list data.
+ */
+struct mcumgr_image_data {
+	/** Image Slot num */
```
*slot

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,195 @@
+	struct img_gr_upload upload;
+	/** Client image list buffer size */
+	int image_list_length;
+	/** Image list buffer*/
```
Add space before */

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,195 @@
+ * @brief IMG mgmt client object.
+ */
+struct img_mgmt_client {
+	/** SMP client object  */
```
There are some double spaces in some comments, replace them with single spaces

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h:
```diff
@@ -0,0 +1,63 @@
+ * @brief OS mgmt client object
+ */
+struct os_mgmt_client {
+	/** SMP client object  */
```
Here too

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -33,6 +33,31 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	default 5
+	help
+	  This define lifetime for SMP client send request. This configure is used if request
+	  with timeout 0 are used..
```
No need for 2 dots

nordicjm reviewed (commented):
A very, very quick flying check, will need to have a proper look later on

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,195 @@
+ */
+struct mcumgr_image_state {
+	/** Status */
+	enum mcumgr_err_t  status;
```
Here too

jheiskan81 commented on subsys/mgmt/mcumgr/transport/src/smp.c:
```diff
@@ -151,8 +168,42 @@ int smp_transport_init(struct smp_transport *smpt)
+	return 0;
+}
+
+struct smp_transport *smp_transport_get(enum smp_transport_type smpt_type)
+{
+#ifdef CONFIG_SMP_CLIENT
+	for (int i = 0; i < registred_client_cnt; i++) {
+		if (registred_smp_clients[i].smpt_type == smpt_type) {
+			return registred_smp_clients[i].smpt;
+		}
+	}
+#endif
+	return NULL;
+}
+
+int smp_transport_register(struct smp_transport *smpt, enum smp_transport_type smpt_type)
```
OK good point. I will extend own Kconfig for transport. Is it fine add support first to serial transport?

jheiskan81 commented on subsys/mgmt/mcumgr/transport/src/smp.c:
```diff
@@ -131,9 +141,16 @@ smp_handle_reqs(struct k_work *work)
 	}
+#ifdef CONFIG_SMP_CLIENT
+	/* Send Client request  */
+	while ((nb = net_buf_get(&smpt->tx_fifo, K_NO_WAIT)) != NULL) {
+		smpt->functions.output(nb);
+	}
```
Then we have one object linked list what transport layer poll and use.

jheiskan81 replied:
Here we need to balance waste of ram or extend code base which one is important.

nordicjm replied:
I'll have a think about how this can be done, might just need to use a bitmask field to specify which modes are enabled, 1 bit for server, 1 bit for client.

As an aside: have you tested using e.g. the UART transport in both server and client modes at the same time? I like the idea that both can be used, but would think that client would need to take priority and block server mode if it is awaiting for something to be returned.

jheiskan81 replied:
I think if I add tx fifo to SMP client structure and use only pointer at smp transport structure that is balanced solution. SMP client init will set pointer and init list.

jheiskan81 replied:
Server and client work fine same time.  Client is blocking request when wait a response but server can be run background. Now transport tx and rx is handled at same worker queue.

jheiskan81 commented on subsys/mgmt/mcumgr/transport/src/smp_udp.c:
```diff
@@ -394,6 +397,9 @@ static void smp_udp_start(void)
 
 	rc = smp_transport_init(&configs.ipv6.smp_transport);
+	if (rc == 0 && IS_ENABLED(CONFIG_SMP_CLIENT)) {
+		rc = smp_transport_register(&configs.ipv6.smp_transport, SMP_UDP_TRANSPORT);
```
OK. If both can be supported same time then I need own Type for IPv4 and 6.

jeppenodgaard commented on subsys/mgmt/mcumgr/smp/src/smp.c:
```diff
@@ -461,7 +472,9 @@ int smp_process_request_packet(struct smp_streamer *streamer, void *vreq)
 
 	smp_free_buf(req, streamer->smpt);
-	smp_free_buf(rsp, streamer->smpt);
+	if (rsp) {
```
Current version of `smp_free_buf` checks if rsp is NULL so the previous implementation seems ok to me. Is there a specific reason to change it now?

jheiskan81 replied:
Ah OK if it safe to call then check is not necessary. Client state never allocated response buffer. I will put that part back of original.

jeppenodgaard reviewed (commented):
Thank you for creating this PR. I am pretty sure I will use this functionality in the future!
I am not that familiar with MCUmgr and SMP. Please have that in mind when reading my comments :smile:

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,611 @@
+	     zcbor_uint32_put(zse, upload_state->image_num) && zcbor_tstr_put_lit(zse, "data") &&
+	     zcbor_bstr_encode_ptr(zse, &temp_data, 1) && zcbor_tstr_put_lit(zse, "len") &&
+	     zcbor_size_put(zse, upload_state->image_size) && zcbor_tstr_put_lit(zse, "off") &&
+	     zcbor_size_put(zse, upload_state->image_size);
```
Yes it make sense. Luckily in this function we calculate just header size but good spot.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,611 @@
+		rc = zcbor_map_decode_bulk(zsd, list_res_decode, ARRAY_SIZE(list_res_decode),
+					   &decoded);
+		if (rc) {
+			if (image_info->image_list_length) {
```
This part of code decode map list. if zcbor_map_decode_bulk() return failure it means that there is no more map available. If we have parsed already valid map data that If just break out from while loop.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,611 @@
+
+static int image_state_res_fn(struct net_buf *nb, void *user_data)
+{
+	zcbor_state_t zsd[CONFIG_MCUMGR_SMP_CBOR_MAX_DECODING_LEVELS + 2];
```
+2 is for backup state purpose.  It was already used at server side so I follow same path.

jeppenodgaard replied:
Ah, I see. I can only find [this single line](https://github.com/zephyrproject-rtos/zephyr/blob/9d6ced38226146071369ebc4bca9e78b8aa4786f/include/zephyr/mgmt/mcumgr/smp/smp.h#L51) which has the `+ 2` and also a comment. I think it will be difficult to figure out that the client code took the logic from there, thus difficult to figure out where the `+ 2` comes from.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,611 @@
+		image_upload_buf->image_upload_offset = SIZE_MAX;
+
+		rc = smp_client_send_cmd(active_client->smp_client, nb, image_upload_res_fn,
+					 &spm_mutex, 15);
```
Erase and upload may take some time more time than normal operation. I will do define for this longer possible timeout.

jheiskan81 commented:
@jeppenodgaard thanks for comment's I have fixed those.

jeppenodgaard commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+			memcpy(image_info->image_list[image_info->image_list_length].hash,
+			       hash.value, IMG_MGMT_HASH_LEN);
+			if (version.len > IMG_MGMT_VER_MAX_STR_LEN) {
+				LOG_WRN("Version truncated len %d -> %.*s", version.len,
```
```suggestion
				LOG_WRN("Version truncated len %d -> %d", version.len,
```

nordicjm commented on subsys/mgmt/mcumgr/smp/src/smp.c:
```diff
@@ -408,24 +409,34 @@ int smp_process_request_packet(struct smp_streamer *streamer, void *vreq)
+			}
+
+			cbor_nb_writer_init(streamer->writer, rsp);
+			/* Process the request payload and build the response. */
```
newline space

nordicjm commented on subsys/mgmt/mcumgr/transport/include/mgmt/mcumgr/transport/smp_internal.h:
```diff
@@ -47,6 +48,16 @@ struct zephyr_smp_transport;
+ * @brief Trig SMP client request packet for transmission.
+ *
+ * @param work	The transport to use to send the corresponding response(s).
+ *
```
empty comment line can go

nordicjm commented on subsys/mgmt/mcumgr/transport/src/smp.c:
```diff
@@ -151,10 +162,39 @@ int smp_transport_init(struct smp_transport *smpt)
+		return 0;
+	}
+
+	if (registred_client_cnt == SMP_CLIENT_TRANPORT_MAX) {
```
typo

nordicjm commented on subsys/mgmt/mcumgr/transport/src/smp.c:
```diff
@@ -151,10 +162,39 @@ int smp_transport_init(struct smp_transport *smpt)
+	if (registred_client_cnt == SMP_CLIENT_TRANPORT_MAX) {
+		return -ENOMEM;
+	}
+	registred_smp_clients[registred_client_cnt].smpt_type = smpt_type;
+	registred_smp_clients[registred_client_cnt].smpt = smpt;
+
+	registred_client_cnt++;
```
typos

nordicjm reviewed (commented):
first commit

jheiskan81 commented on include/zephyr/mgmt/mcumgr/transport/smp.h:
```diff
@@ -134,6 +134,22 @@ struct smp_transport {
+	SMP_UDP_IPV6_TRANSPORT
+};
+
+#define SMP_CLIENT_TRANPORT_MAX 4
```
It is define how many Transport can be register for client.

nordicjm replied:
So how is this usable to anyone with an out of zephyr tree transport?

jheiskan81 replied:
Ah OK i understand it is not used outside. Let me move this to internal usage only.

nordicjm replied:
MCUmgr is extensible, anyone can add a custom transport and that has worked since MCUmgr was added to zephyr, if you are saying there is a limitation of the client not being able to support that then we need to revisit if this client can be added at all because that is a huge limitation and hiding it/making it internal usage only does not resolve that issue. A user should be able to add their own transport, and use it in client mode if they so wish.

jheiskan81 replied:
Maybe I to transport register by linked list. So Transport define locally structure which is added to linked list. Then Client side is not limit anything.

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,320 @@
+	smpt = smp_client->smpt;
+
+	while ((nb = net_buf_get(&smp_client->tx_fifo, K_NO_WAIT)) != NULL) {
+		smpt->functions.output(nb);
```
Packet is released at output call. But return value check is needed for error handling response.

jheiskan81 replied:
SMP client will try retry if a packet send so I don't see any reason to change a code only if we want log error code.

jheiskan81 replied:
Request will timeout in the end and buffer is released finally.

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/smp.c:
```diff
@@ -408,24 +409,34 @@ int smp_process_request_packet(struct smp_streamer *streamer, void *vreq)
-		if (rc != 0) {
-			break;
+		if (req_hdr.nh_op == MGMT_OP_READ || req_hdr.nh_op == MGMT_OP_WRITE) {
+			cbor_nb_reader_init(streamer->reader, req);
```
So it is not problem if Client is not need that init?

jheiskan81 replied:
Ok I got your point

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,293 @@
+#define LOG_MODULE_NAME smp_client
+LOG_MODULE_REGISTER(smp_client);
+
+#define SMP_CMD_MAX 4
```
Ok I will add Kconfig for this.

jheiskan81 commented:
@nordicjm Thanks for comments. I fix typos and change request. I check possible output() return error and it not affect a problem because client retry logic try again and finally timeout stop request handling and free memory.

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,320 @@
+{
+	/* Pre config SMP header structure */
+	memset(header, 0, sizeof(struct smp_hdr));
+	header->nh_version = SMP_MCUMGR_VERSION_1;
```
I change to version 2

jheiskan81 replied:
I leave version 1 for still use because of Mcuboot still use that.

jheiskan81 commented:
@nordicjm @de-nordic Please review.

jheiskan81 commented:
reabased with upstream

jheiskan81 commented:
Fixed SMP version back to 1 that we keep support to Mcuboot

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+			goto out;
+		}
+		/* Check that mandatory parameters have decoded */
+		if (hash.len != IMG_MGMT_HASH_LEN || !version.len || slot_num == -1) {
```
use the zcbor found check for slot num

jheiskan81 replied:
what is this found check?

jheiskan81 replied:
Ah thanks. I will use this function.

nordicjm commented on include/zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h:
```diff
@@ -0,0 +1,195 @@
+ */
+struct mcumgr_image_data {
+	/** Image slot num */
+	int slot_num;
+	/** Image number */
+	int img_num;
```
uint

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+	uint8_t temp_buf[MCUMGR_UPLOAD_INIT_HEADER_BUF_SIZE];
+	uint8_t temp_data;
+
+	/* Calculating message over head by data legnht 1 */
```
*of
*header
*by = with
*length of

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt_client/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,166 @@
+
+static struct os_mgmt_client *active_client;
+static K_SEM_DEFINE(spm_mutex, 0, 1);
+static K_MUTEX_DEFINE(mcumgr_mutex);
```
it might be static but rename to fit the style of file e.g. mcmmgr_os_client_grp_mutex or similar

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+			     zcbor_size_put(zse, active_client->upload.offset);
+		}
+		/* Write hash when it defined and offset is 0 */
+		if (ok && !active_client->upload.offset && active_client->upload.hash_initialized) {
+			ok = zcbor_tstr_put_lit(zse, "sha") &&
+			     zcbor_bstr_encode_ptr(zse, active_client->upload.sha256,
+						   IMG_MGMT_HASH_LEN);
```
Move inside the above if block and then only need to check if okay and if hash is set instead of 3 conditions

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt_client/src/os_mgmt_client.c:
```diff
@@ -0,0 +1,157 @@
+static K_SEM_DEFINE(spm_mutex, 0, 1);
+static K_MUTEX_DEFINE(mcumgr_mutex);
+
+static int reset_res_fn(struct net_buf *nb, void *user_data)
```
this is not resolved

jheiskan81 replied:
OK I add Kconfig.

nordicjm commented on tests/subsys/mgmt/mcumgr/mcumgr_client/src/img_gr_stub.c:
```diff
@@ -0,0 +1,291 @@
+
+	if (sha.len) {
+		if (memcmp(sha.value, image_hash_ptr, 32)) {
+			printf("HASh not same\r\n");
```
fix case

nordicjm commented on tests/subsys/mgmt/mcumgr/mcumgr_client/src/main.c:
```diff
@@ -0,0 +1,295 @@
+	zassert_equal(MGMT_ERR_EOK, rc, "Expected to receive %d response %d\n", MGMT_ERR_EOK, rc);
+	/* Start upload  and test Timeout */
+	rc = img_mgmt_client_upload(&client, image_dummy, 1024, &response);
+	zassert_equal(MGMT_ERR_ENOMEM, rc, "Expected to receive %d response %d\n", MGMT_ERR_ENOMEM,
```
remove \n from all zassert calls, it is not needed

nordicjm commented on tests/subsys/mgmt/mcumgr/smp_client/src/main.c:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	for (int i = 0; i < 4; i++) {
+
```
empty newline can go

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,317 @@
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&smp_client_data.cmd_list, entry, tmp, node) {
+		if (k_uptime_get() < entry->timestamp) {
+			time_stamp_cmp = entry->timestamp - k_uptime_get();
+			if (time_stamp_cmp < CONFIG_SMP_CMD_RETRY_TIME &&
```
Ah good tip

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,317 @@
+		if (cb) {
+			cb(nb, user_data);
+		}
+	}
```
Yes I will fix that end of function return error.

jheiskan81 commented on subsys/mgmt/mcumgr/smp/src/client.c:
```diff
@@ -0,0 +1,317 @@
+	if (nb->len > sizeof(smp_header)) {
+		/* Update Length */
+		smp_header.nh_len = sys_cpu_to_be16(nb->len - sizeof(smp_header));
+		smp_header.nh_group = sys_cpu_to_be16(smp_header.nh_group),
+		memcpy(nb->data, &smp_header, sizeof(smp_header));
```
Buffer allocate set buffer length to header size. If user have add data that need to be updated for smp header. If length is same than header size that means that payload length is zero.

jheiskan81 commented on include/zephyr/mgmt/mcumgr/grp/os_mgmt/os_mgmt_client.h:
```diff
@@ -0,0 +1,63 @@
+extern "C" {
+#endif
+
+struct smp_client_object;
```
I think this is valid way to do it. Of course I can add include.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+
+#define MCUMGR_UPLOAD_INIT_HEADER_BUF_SIZE 128
+
+#define MCUMGR_MGMT_FLASH_OPERATION_TIMEOUT 15
```
OK I will do Kconfig. I'm not a big fan to add too much those but I think this is good to add.

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+		ZCBOR_MAP_DECODE_KEY_DECODER("hash", zcbor_bstr_decode, &hash),
+		ZCBOR_MAP_DECODE_KEY_DECODER("slot", zcbor_int32_decode, &slot_num),
+		/* Optional */
+		ZCBOR_MAP_DECODE_KEY_DECODER("image", zcbor_int32_decode, &img_num),
```
I still see server side int32

jheiskan81 replied:
img_mgmt_state_read() still use int32

nordicjm replied:
It does not https://github.com/zephyrproject-rtos/zephyr/blob/83b074c4185d2ddbae3a2936c02c572c96052222/subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c#L258
And in mcuboot it never did: https://github.com/mcu-tools/mcuboot/blob/main/boot/boot_serial/src/boot_serial.c#L324

jheiskan81 replied:
Ok

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+		return 0;
+	}
+	cbor_length = zse->payload - temp_buf;
+	/* Return Message over length */
```
header length

jheiskan81 commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+		}
+		/* Write offset */
+		if (ok) {
+			ok = zcbor_tstr_put_lit(zse, "off") &&
+			     zcbor_size_put(zse, active_client->upload.offset);
```
Yes true I will move.

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+
+		if (offset_before_send + write_length < active_client->upload.offset) {
+			/* Offset further than expected which indicate upload session resume */
+			goto end;
```
Network client might get an 8KB chunk, might save it here, it might get interrupted, it reconnects and gets a 32KB chunk, writes the first packet and it hits this part and exits until it rights the next part... What happens to the extra 32-8 = 24KB of data from this packet which was not written?

jheiskan81 replied:
OK. In that case we need to download same data again based on server reported offset. Coap use 512 bytes and http 1024 for requesting data.

nordicjm replied:
Is this always the case or can it change, i.e. if you are one cellular network and you disconnect then connect to another, could the size of the data you receive or get this callback on be different (or does the web client buffer it)?

jheiskan81 replied:
Download client config received data size to request.

jheiskan81 commented on tests/subsys/mgmt/mcumgr/mcumgr_client/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+
+target_sources(app PRIVATE ${app_sources})
+
+add_compile_definitions(CONFIG_SMP_CLIENT=y)
```
We have been used same method when we want stub this module. If I add this to prj.conf It will include client src files.

nordicjm replied:
If you need this for stubs then why does the client Kconfig files have the ztest part in? If so, remove the ztest part from those and manually add the source files for compilation in this cmake file

jheiskan81 replied:
I have added configs to prj.conf

nordicjm commented on subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c:
```diff
@@ -0,0 +1,617 @@
+static const char smp_images_str[] = "images";
+#define IMAGES_STR_LEN (sizeof(smp_images_str) - 1)
+
+static void reset_image_map_decoder_list(struct zcbor_map_decode_key_val *map, size_t map_size)
```
Replace this by using this https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/mcumgr/util/src/zcbor_bulk.c#L105

jheiskan81 replied:
Very good point I will fix this.

jheiskan81 commented on include/zephyr/mgmt/mcumgr/transport/smp.h:
```diff
@@ -134,6 +134,31 @@ struct smp_transport {
+	/** SMP UDP IPv4 */
+	SMP_UDP_IPV4_TRANSPORT,
+	/** SMP UDP IPv6 */
+	SMP_UDP_IPV6_TRANSPORT
```
OK I will extend this and extend API for support that.

nordicjm commented on subsys/mgmt/mcumgr/Kconfig:
```diff
@@ -33,6 +33,37 @@ config MCUMGR_SMP_LEGACY_RC_BEHAVIOUR
+	range 2 30
+	default 5
+	help
+	  This define lifetime for SMP client send request. This configure is used if request
+	  with timeout 0 are used.
```
```suggestion
	  This define lifetime for SMP client send request. This configure is used if a request
	  with a timeout of 0 is used.
```

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt_client/Kconfig:
```diff
@@ -0,0 +1,29 @@
+if MCUMGR_GRP_OS_CLIENT
+
+	config MCUMGR_GRP_OS_CLIENT_ECHO
+	bool "Support for echo command request"
+
+	config MCUMGR_GRP_OS_CLIENT_RESET
+	bool "support for reset command request"
```
These can (optionally) be defaulted to y

nordicjm commented on subsys/mgmt/mcumgr/transport/src/smp_uart.c:
```diff
@@ -101,6 +104,11 @@ static int smp_uart_init(void)
 	if (rc == 0) {
 		uart_mcumgr_register(smp_uart_rx_frag);
+#ifdef CONFIG_SMP_CLIENT
+		smp_client_transport.smpt = &smp_uart_transport;
+		smp_client_transport.smpt_type = SMP_SERIAL_TRANSPORT;
+		smp_client_transport_register(&smp_client_transport);
```
Missing change in smp_shell.c

nordicjm reviewed (changes requested):
Isssues found in test:

- Client must not respond to packets with error responses, it needs to ignore corrupt packets and handle them i.e. with retransmissions
- Client is not removing the data from the packet and it then tries to look at an invalid packet
- Upload does not set the image status to 0, therefore it always fails because it always has the default value of EINVAL

nordicjm commented:
Also logging does not work, instead of having:
```
#define LOG_MODULE_NAME smp_client
LOG_MODULE_REGISTER(smp_client);
```
Use:
```
LOG_MODULE_REGISTER(mcumgr_client, <log_level_here_from_Kconfig>);
```

jheiskan81 commented:
@nordicjm Thanks for testing and reporting issues. I have fixed issues and also fix logging.

jheiskan81 commented:
Fixed Compliance checks

nordicjm commented on subsys/mgmt/mcumgr/grp/os_mgmt_client/Kconfig:
```diff
@@ -0,0 +1,35 @@
+if MCUMGR_GRP_OS_CLIENT
+
+	config MCUMGR_GRP_OS_CLIENT_ECHO
+	default y
```
No indenting in Kconfig files for if... blocks

nordicjm commented on subsys/mgmt/mcumgr/smp_client/src/client.c:
```diff
@@ -0,0 +1,326 @@
+	}
+
+	smp_read_hdr(nb, &smp_header);
+	if (nb->len > sizeof(smp_header)) {
```
Would flip this to error out if there is not at least `sizeof(smp_header)` to send, as that would not be valid, then the part below can be part of the main flow as it will only send packets of a minimum valid length

nordicjm reviewed (changes requested):
2 minor nits then ready for approval, thanks!

jheiskan81 commented:
@nordicjm Thanks for review. I fixed those raised case.

de-nordic commented on subsys/mgmt/mcumgr/transport/src/smp.c:
```diff
@@ -155,6 +160,33 @@ int smp_transport_init(struct smp_transport *smpt)
+void smp_client_transport_register(struct smp_client_transport_entry *entry)
+{
+	if (smp_client_transport_get(entry->smpt_type)) {
+		/* Already in list */
```
We should probably have some warning here, as a LOG_DBG, as this would be strange behavior when certain transport type is registered twice; this also checks type only so it does not distinguish between different handler being registered for the same type.

jheiskan81 replied:
I can add that no problem.

de-nordic reviewed (approved):
Looks OK, thanks!

kartben commented:
Is there a plan to add documentation and/or a sample for this? It would really help with making sure people are aware this now exists. Apologies if there is already an open issue or PR for this and I just missed it.
